src/QsCompiler/SyntaxProcessor/TypeInference/InferenceContext.fsi

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Quantum.QsCompiler.SyntaxProcessing.TypeInference
+
+open Microsoft.Quantum.QsCompiler.DataTypes
+open Microsoft.Quantum.QsCompiler.SyntaxProcessing
+open Microsoft.Quantum.QsCompiler.SyntaxTree
+open Microsoft.Quantum.QsCompiler.Transformations.Core
+
+/// The inference context is an implementation of Hindley-Milner type inference. It is a source of fresh type parameters
+/// and can unify types containing them.
+type InferenceContext =
+    /// <summary>
+    /// Creates a new inference context using the given <see cref="SymbolTracker"/>.
+    /// </summary>
+    new: symbolTracker:SymbolTracker -> InferenceContext
+
+    /// Diagnostics for all type variables that are missing substitutions.
+    member AmbiguousDiagnostics: QsCompilerDiagnostic list
+
+    /// Updates the position of the statement in which types are currently being inferred.
+    member UseStatementPosition: position:Position -> unit
+
+    /// <summary>
+    /// Creates a fresh type parameter originating from the given <paramref name="source"/> range.
+    /// </summary>
+    member internal Fresh: source:Range -> ResolvedType
+
+    /// <summary>
+    /// Unifies the <paramref name="expected"/> type with the <paramref name="actual"/> type.
+    /// </summary>
+    member internal Unify: expected:ResolvedType * actual:ResolvedType -> QsCompilerDiagnostic list
```

Comment:
Hm, what confuses me a bit is why you don't have a method for unification that computes the more general type of the two, and one that computes the less general type of the two ("covariant vs contravariant unification"). How does that work currently?

Reply:
That is `UnifyByOrdering`.

https://github.com/microsoft/qsharp-compiler/blob/70f506f9ad6b14554572bcdccba4485a4daf821c/src/QsCompiler/SyntaxProcessor/TypeInference/InferenceContext.fs#L362-L366

Reply:
I updated the doc comments for `Unify` and `Intersect`. For `Unify`, `actual` must be a subtype of `expected`. `Intersect` returns a type that is a supertype of the given types.

The reason that the more general functions are private is simply because there is no reason for any other part of the compiler to ask for an invariant or contravariant unification or intersection. That only happens when recursing into an array or callable type, which only occurs in the inference module itself.